### PR TITLE
Adds archetype downloads + installation.

### LINKED ISSUE #2: Feature: Install from various sources
* [x] Do actual CLI with arguments and flags (share with `builder`?)
* [x] From NPM (at a version?)
* [x] From filesystem
* [x] From GitHub / generic git (at a version?)
* [x] Maybe do GH `ORG/PROJECT` shortcut if possible (?) (Like use `npm` internals)
* [x] Hook up to Travis CI to actually init an archetype and run the new project's CI commands.

Optional:

* [x] `builder-init help` #5 
* [ ] `init.js:scripts` #12

### LINKED ISSUE #5: Feature: `builder-init help`

### LINKED ISSUE #6: Feature: Documentation / usage
* [x] Basic usage section
* [x] File layout: `init.js`, `init/`
* [x] Example: Init from local file.
* [x] Example: Init from github.
* [x] Example: Init from npm.
* [x] Note: Destination directories cannot already exist.

### PR DESCRIPTION
**Pull request by ryan-roemer**
At long last, `builder-init` is (hopefully) ready for prime time!
### Overview

We use a process spawn of `npm pack` to do the heavy lifting for downloading
archetype tarballs from file, npm, git, etc. We then expand the tarballs in
a temp directory and process them into an outputted project.
- Downloads + installs archetype templates from npm, git, file, etc. Fixes #2
- Adds `builder-init --help`. Fixes #5
- Adds `builder-init --version`.
- Adds `builder-init <archetype> --prompts=JSON` helper for automation.
- Adds usage / documenation. Fixes #6
- Add near-functional tests for most full user interaction flows.
### Kicking the Tires

This should work:

``` sh
$ builder-init FormidableLabs/builder-react-component#chore-init-install
```

from the parallel PR https://github.com/FormidableLabs/builder-react-component/pull/33
### Usage

```
$ builder-init -h
Usage:

  builder-init [flags] <archetype>

Flags:

  --help
  --version

Examples: 

`builder-init` can install templates from any source that `npm` can, e.g.:

  (npm)    builder-init builder-react-component
  (npm)    builder-init builder-react-component@0.2.0
  (github) builder-init FormidableLabs/builder-react-component
  (github) builder-init FormidableLabs/builder-react-component#v0.2.0
  (git)    builder-init git+ssh://git@github.com:FormidableLabs/builder-react-component.git
  (git)    builder-init git+ssh://git@github.com:FormidableLabs/builder-react-component.git#v0.2.0
  (file)   builder-init /FULL/PATH/TO/builder-react-component
```
### Sample Flow

```
$ builder-init FormidableLabs/builder-react-component#chore-init-install
builder-react-component-0.1.4.tgz
[builder-init] Preparing templates for: builder-react-component
? Package / GitHub project name (e.g., 'whiz-bang-component') green-button
? GitHub organization name (e.g., 'AcmeCorp') GreenCorp
? Package description The Green Button!
? License date 2016
? License organization (e.g., you or your company) GreenCorp
? Destination directory to write green-button

[builder-init] Wrote files:
 - green-button/.babelrc
 - green-button/.builderrc
 - green-button/.editorconfig
 - green-button/.travis.yml
 - green-button/CONTRIBUTING.md
 - green-button/LICENSE.txt
 - green-button/README.md
 - green-button/package.json
 - green-button/.gitignore
 - green-button/.npmignore
 - green-button/demo/app.jsx
 - green-button/demo/index.html
 - green-button/src/index.js
 - green-button/src/components/green-button.jsx
 - green-button/test/client/main.js
 - green-button/test/client/test.html
 - green-button/test/client/spec/components/green-button.spec.jsx

[builder-init] New builder-react-component project is ready at: green-button
```
### Review

Admittedly, this PR is a biggie, so appreciate any review we can get here.
However, the bulk of the line changes are in the tests, and we've now got
pretty comprehensive coverage of some awkward corner cases that can arise.

/cc @coopy @benbayard @chaseadamsio @exogen @boygirl @zachhale @kenwheeler @jstrimpel @baer

### CODE REVIEW

**Comment by ryan-roemer** [comment]
I've reproduced and am checking out node 4+ CI failures.

**Comment by ryan-roemer** [comment]
**Coveralls Note**: CI decrease is acceptable as it's user prompts for `destination` that we normally mock out anyways. I'll leave the red lines for now and we can add coverage / tickets if needed later.

File: README.md [hunk 1]
```diff
@@ -20,13 +21,92 @@ from scratch, so you have to start somewhere...
...
 ```
-TODO: Add usage, documentation.
-https://github.com/FormidableLabs/builder-init/issues/6
+
+### Automating Prompts
+
+To facilitate automation, notably testing an archetype by generating a project
+with `builder-init` and running the project's tests as part of CI, there is a
+special `--prompts=JSON_OBJECT` flag that skips the actual input prompts and
+injects fields straight from a JSON object.
```
Comment (coopy) [inline README.md]: This is a great feature!

File: README.md [hunk 2]
```diff
@@ -20,13 +21,92 @@ from scratch, so you have to start somewhere...
...
+[`builder-react-component/.travis.yml`](https://github.com/FormidableLabs/builder-react-component/blob/master/.travis.yml)
+which initializes the archetype's templates for a fresh project with canned
+`--prompts` values, npm installs dependencies, then runs the same `builder`
+tasks used in the project's CI.
+
+
+## Archetype Templates
 
-## Templates
+Authoring a templates for an archetype consists of adding the following to your
```
Comment (coopy) [inline README.md]: drop `a`: `Authoring templates ...`

File: test/server/spec/bin/builder-init.spec.js [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+      return fn(_done);
+    } catch (err) {
+      return _done(err);
+    }
+  };
+};
+
+// Mock key I/O parts of the flow.
+/*eslint-disable max-statements*/
+var mockFlow = function (extracted, root) {
```
Comment (coopy) [inline test/server/spec/bin/builder-init.spec.js]: Holy mocking, Batman!

File: README.md (hunk 1 above)
Comment (rgerstenberger) [inline README.md]: Where is that `JSON_OBJECT` defined? If it's local to an archetype, how would one set that up?

File: README.md [hunk 4]
```diff
@@ -20,13 +21,92 @@ from scratch, so you have to start somewhere...
...
+archetype source:
```
Comment (rgerstenberger) [inline README.md]: I'd add a bit more about templates here. It's unclear from looking at the examples, how folder structure is handled, what type of templating is allowed, etc.

File: README.md [hunk 5]
```diff
@@ -20,13 +21,92 @@ from scratch, so you have to start somewhere...
...
+
+* **`init.js`**: A control file for user prompts and data. See, e.g.,
```
Comment (rgerstenberger) [inline README.md]: This is also unclear, please add more info on `init.js` and `prompts, destination, derived`. What parts are required, what are valid options, invalid values, how are they used, etc.

Comment (rgerstenberger) [inline README.md]: Nevermind, missed the pieces down below.

**Comment by rgerstenberger** [comment]
Aside from some confusion on the docs, this looks good. I'd like another reviewer to look over the `process` and `fs` code chunks in case I missed a potential problem deep in node.

File: README.md (hunk 1 above)
Comment (ryan-roemer) [inline README.md]: @rgerstenberger -- Does the example three lines down on L86-87 help answer that question?

File: README.md (hunk 2 above)
Comment (ryan-roemer) [inline README.md]: Yup.

File: README.md (hunk 4 above)
Comment (ryan-roemer) [inline README.md]: Added blurb we talked about.

File: test/server/spec/bin/builder-init.spec.js (hunk 3 above)
Comment (ryan-roemer) [inline test/server/spec/bin/builder-init.spec.js]: I know!

The cool thing is that `mock-fs` has been wonderful for having a ton of "file-based" tests that don't actually touch any filesystem so that I can just define them with objects in place in the tests that consume them.

**Comment by coopy** [comment]
This is a metric ton of good work! Tested out the binary, worked like a charm.
:+1: :100: :shipit:

**Comment by ryan-roemer** [comment]
@rgerstenberger @coopy -- Thanks for the review! Rebasing and going for the merge!

Everyone else -- post-merge review welcome!

**Comment by ryan-roemer** [comment]
**Released as `builder-init@0.0.1`!**
